### PR TITLE
Recover optional fields for 4.5 jobs

### PR DESCRIPTION
**Pull request by hongkailiu**
https://issues.redhat.com/browse/DPTP-1288

```bash
python3 ./hack/recover-4.5.py
ci-operator-prowgen --from-dir ./ci-operator/config/ --to-dir ./ci-operator/jobs
sanitize-prow-jobs --prow-jobs-dir=./ci-operator/jobs --config-path ./core-services/sanitize-prow-jobs/_config.yaml
```

with `./hack/recover-4.5.py`: 

```python
#!/usr/bin/env python3

import argparse
import logging
import os
import yaml


def getJobFromMaster(file_path, job):
    if not os.path.isfile(file_path):
        return None
    context = job.get('context', '')
    if not context:
        return None
    with open(file_path) as file:
        data = yaml.safe_load(file)
        for job_type in ["presubmits"]:
            for repo in data.get(job_type, {}):
                for j in data.get(job_type, {}).get(repo, []):
                    if j.get('context', '') == context:
                        return j
    return None


def ensure(job_dir):
    for dirpath, _, filenames in os.walk(job_dir):
        for filename in filenames:
            if not filename.endswith('-presubmits.yaml') or not "release-4.5" in filename:
                continue
            file_path = os.path.join(dirpath, filename)
            with open(file_path) as file:
                data = yaml.safe_load(file)
                for job_type in ["presubmits"]:
                    for repo in data.get(job_type, {}):
                        jobs = []
                        for job in data.get(job_type, {}).get(repo, []):
                            job_from_master = getJobFromMaster(file_path.replace("release-4.5", "master"), job)
                            if not job_from_master is None:
                                job['always_run'] = job_from_master.get('always_run', False)
                                job['run_if_changed'] = job_from_master.get('run_if_changed', '')
                                job['optional'] = job_from_master.get('optional', False)
                                job['max_concurrency'] = job_from_master.get('max_concurrency', 0)
                                job['skip_report'] = job_from_master.get('skip_report', False)
                            jobs.append(job)
                        data[job_type][repo] = jobs
            with open(file_path, 'w') as file:
                yaml.dump(data, file)


logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s:%(levelname)s:%(message)s"
    )


def main():
    """main function."""
    logging.info("checking jobs ...")
    parser = argparse.ArgumentParser(description='Ensure prow jobs\' cluster.')
    parser.add_argument('-d', '--job-dir', default='./ci-operator/jobs',
                        help="the path to the job directory")
    args = parser.parse_args()
    ensure(args.job_dir)



if __name__ == "__main__":
    main()
```

/cc @petr-muller @wking

### CODE REVIEW

**Comment by petr-muller** [comment]
`job['always_run'] = job_from_master.get('always_run', False)`

I think the "default if not present" should be `True` here (jobs are `always_run: true` by default)

File: ci-operator/jobs/openshift/installer/openshift-installer-release-4.5-presubmits.yaml [hunk 1]
```diff
@@ -588,7 +591,6 @@ presubmits:
       ci-operator.openshift.io/prowgen-controlled: "true"
       pj-rehearse.openshift.io/can-be-rehearsed: "true"
     name: pull-ci-openshift-installer-release-4.5-e2e-gcp-upgrade
-    optional: true
```
Comment (petr-muller) [inline ci-operator/jobs/openshift/installer/openshift-installer-release-4.5-presubmits.yaml:591]: Can we confirm that this one is really desired to be *not* optional?

File: ci-operator/jobs/openshift/ovn-kubernetes/openshift-ovn-kubernetes-release-4.5-presubmits.yaml [hunk 2]
```diff
@@ -103,7 +103,6 @@ presubmits:
     labels:
       pj-rehearse.openshift.io/can-be-rehearsed: "true"
     name: pull-ci-openshift-ovn-kubernetes-release-4.5-e2e-gcp-ovn
-    optional: true
```
Comment (petr-muller) [inline ci-operator/jobs/openshift/ovn-kubernetes/openshift-ovn-kubernetes-release-4.5-presubmits.yaml:106]: This too.

File: ci-operator/jobs/openshift/installer/openshift-installer-release-4.5-presubmits.yaml (hunk 1 above)
Comment (hongkailiu) [reply ci-operator/jobs/openshift/installer/openshift-installer-release-4.5-presubmits.yaml:591]: Checked on master: It is NOT optional

https://github.com/openshift/release/blob/5d8e151961b0aa3b1c6cf245c581ab0597392d42/ci-operator/jobs/openshift/installer/openshift-installer-master-presubmits.yaml#L1090-L1153

File: ci-operator/jobs/openshift/ovn-kubernetes/openshift-ovn-kubernetes-release-4.5-presubmits.yaml (hunk 2 above)
Comment (wking) [inline ci-operator/jobs/openshift/ovn-kubernetes/openshift-ovn-kubernetes-release-4.5-presubmits.yaml:106]: dropping `optional` is surprising to me.  @JacobTanenbaum , @squeed ,  can you speak to the indended behavior of e2e-gcp-ovn in 4.5?

File: ci-operator/jobs/openshift/installer/openshift-installer-release-4.5-presubmits.yaml (hunk 1 above)
Comment (wking) [inline ci-operator/jobs/openshift/installer/openshift-installer-release-4.5-presubmits.yaml:591]: Dropping optional is surprising to me.  @abhinavdahiya , can you speak to the intended behavior of e2e-gcp-upgrade in 4.5?

File: ci-operator/jobs/openshift/ovn-kubernetes/openshift-ovn-kubernetes-release-4.5-presubmits.yaml (hunk 2 above)
Comment (hongkailiu) [reply ci-operator/jobs/openshift/ovn-kubernetes/openshift-ovn-kubernetes-release-4.5-presubmits.yaml:106]: Checked this too.

https://github.com/openshift/release/blob/5d8e151961b0aa3b1c6cf245c581ab0597392d42/ci-operator/jobs/openshift/ovn-kubernetes/openshift-ovn-kubernetes-master-presubmits.yaml#L158-L250

**Comment by hongkailiu** [comment]
> `job['always_run'] = job_from_master.get('always_run', False)`
> 
> I think the "default if not present" should be `True` here (jobs are `always_run: true` by default)

Checked each change in my PR.
We always set the field `always_run` explicitly, no matter it is `true` or `false`.

File: ci-operator/jobs/openshift/ovn-kubernetes/openshift-ovn-kubernetes-release-4.5-presubmits.yaml (hunk 2 above)
Comment (squeed) [reply ci-operator/jobs/openshift/ovn-kubernetes/openshift-ovn-kubernetes-release-4.5-presubmits.yaml:106]: not sure what the best practice is; we also run e2e-aws, which is not optional.

File: ci-operator/jobs/openshift/installer/openshift-installer-release-4.5-presubmits.yaml (hunk 1 above)
Comment (petr-muller) [reply ci-operator/jobs/openshift/installer/openshift-installer-release-4.5-presubmits.yaml:591]: What I mean is: apparently somebody manually set this on the release branch, I don't think we should overwrite it.

Comment (petr-muller) [reply ci-operator/jobs/openshift/installer/openshift-installer-release-4.5-presubmits.yaml:591]: I understand it is not optional on master, but that does not mean it should also be not-optional on release branch.

Comment (hongkailiu) [reply ci-operator/jobs/openshift/installer/openshift-installer-release-4.5-presubmits.yaml]: Got it.
Keep it unchanged.

**Comment by hongkailiu** [comment]
/hold

Let us wait for the developers confirmation if we should include the 2nd commit about the 2 controversial `optional` fields.

**Pull request merged**